Improve spectral classes

jonathanjfshaw opened the pull request:
This PR makes various inter-related fixes and improvements (probably too many) to the spectral classes:

* Fix #8 for FFT (but not for Welch yet)
* Allow FFT nfft to vary dynamically dependent on input length
* Allow FFT rate to come from meta
* Rename FFT 'fs' parameter to 'rate' for consistency with Welch and rest of timeflux
* Fix bug in FFT where one-sided (the default) was not working because pd.Series has no attribute 'real' in pd.apply
* Don't bother converting dataframe to real numbers because https://docs.scipy.org/doc/numpy/reference/generated/numpy.fft.rfft.html#numpy.fft.rfft) says: "If the input a contains an imaginary part, it is silently discarded."
 * Use same dimension "frequency" in FFT, Welch and Bands so that Bands can come after an FFT.
* Convert Bands output to amplitude real number.
* Allow FFT to interpolate NaNs, so that a whole 20 second window is not discarded because of a single NaN
* Document relative parameter for Bands

jonathanjfshaw committed:
Improve spectral

jonathanjfshaw committed:
Fix int nfft param

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -63,36 +64,50 @@ def __init__(self, fs=1.0, nfft=None, return_onesided=True):
+            self._interpolation = {**self._interpolation, **interpolation}
+        else:
+            self._interpolate = interpolation
+
```
I prefer the user to use an interpolation node before the FFT if needed, since it's not the role of the FFT Node to bother with interpolation and for DRY purposes.

jonathanjfshaw replied:
Agreed, removed

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -104,17 +119,26 @@ def update(self):
+            if 'rate' in self.i.meta:
+                self._rate = self.i.meta['rate']
+            else:
+                self._rate = 1.0
```
We could add warning log in that case.

jonathanjfshaw replied:
Done

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -104,17 +119,26 @@ def update(self):
+        # FFT cannot handle NaNs and will return NaN if a single input value is NaN.
+        if self._interpolate:
+            self.o.data = self.o.data.interpolate(**self._interpolation)
+
```
See my previous comment.

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -63,36 +64,50 @@ def __init__(self, fs=1.0, nfft=None, return_onesided=True):
+            raise ValueError('nfft must be less than or equal to length of chunk.')
         else:
-            self._nfft = int(self._nfft)
+            self.o.data = self.o.data.tail(self._fixed_nfft)
```
I don't like tailing the data. If such a thing happen, the user should be at least warned.

jonathanjfshaw replied:
The performance with FFT is better if you keep to powers of 2 like 64, 128, 256.  https://stackoverflow.com/questions/46357589/numpy-fft-is-fast-for-lengths-that-are-products-of-small-primes-but-how-small But it's not simple to provide a window that delivers exactly 256 samples

jonathanjfshaw replied:
It seems easier and safer to get the window to give a sensible round number of seconds like 1.5 or 2, then getting the FFT to truncate to 256 samples, rather than calculating that your window needs to be 1.234678 seconds long and trusting the Window node to give you the result you expect.

jonathanjfshaw replied:
We could of course add a max_samples argument to the Window node and give it responsibiity for this.

bertrandlalo replied:
I understand. I have a node for that purpose, `FixedWindow`, where you specify length with samples instead of seconds. I'll push it tomorrow in module `timeflux.nodes.window`. 
Here, tailing will make the data discontinuous. Imagine on a sinus signal!

jonathanjfshaw replied:
OK, removed

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -63,36 +64,50 @@ def __init__(self, fs=1.0, nfft=None, return_onesided=True):
+        self._fixed_rate = rate
+        self._rate = rate
+        self._nfft = None if nfft is None else int(nfft)
+        self._fixed_nfft = self._nfft
```
See my main comment. I'm not sure this is relevant to allow for a changing nfft.

jonathanjfshaw replied:
Agreed, removed.

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -277,7 +303,8 @@ def update(self):
-
             band['port'].data = pd.DataFrame(columns=self.i.data.space.values, index=self.i.data.time.values,
                                              data=band_power)
+            if self._amplitude:
```
Usually, the input signals is a power, so it's already a real values. 
Instead of adding an argument, I'd prefer the user to add a node just after to apply this `abs` operation, insofar as this is really case dependant, and one could also want to take the square, sqrt, the angle, or else...
What is your use-case ?

jonathanjfshaw replied:
I had a sensible real valued signal (mean zero, sd 1), I passed it into the FFT, I passed that into Bands, I got out a complex number I wasn't expecting and didn't really understand how to interpret. Googling stack exchange suggested I just needed to abs it to get what I wanted, and I figured this was probably the 95% use case for other people.

bertrandlalo replied:
This node may also be used after a Welch node, where you need to average powers in a frequency band. 
For your use-case, I would definitely do a Branch with 3 nodes: FixedWindow(length=2^N) -> FFT -> Abs(on ports)

jonathanjfshaw replied:
OK, removed

bertrandlalo reviewed (changes requested):
Thank. you very much for the review and PR. 
I should comment the list of your changes: 

> Fix #8 for FFT (but not for Welch yet)

OK ! 👍 

>  Allow FFT nfft to vary dynamically dependent on input length

_Why_ would you want such a thing? To me, this node should be used as part of a Branch, with a Window node before to ensure that the chunks are of same length. Actually, we could even warn the user if this is not the case. The number of samples decides the frequency axis, and you don't want it to be changed. 

> Allow FFT rate to come from meta

OK ! 👍 

> Rename FFT 'fs' parameter to 'rate' for consistency with Welch and rest of timeflux

OK ! 👍 

> Fix bug in FFT where one-sided (the default) was not working because pd.Series has no attribute 'real' in pd.apply

I don't understand your point. Is there a related issue ? 

> Don't bother converting dataframe to real numbers because https://docs.scipy.org/doc/numpy/reference/generated/numpy.fft.rfft.html#numpy.fft.rfft) says: "If the input a contains an imaginary part, it is silently discarded."

I don't understand your point ? 

> Use same dimension "frequency" in FFT, Welch and Bands so that Bands can come after an FFT.
Convert Bands output to amplitude real number.

OK ! 👍 

> Allow FFT to interpolate NaNs, so that a whole 20 second window is not discarded because of a single NaN

NOK ! I prefer the user to use an interpolation node before the FFT if needed, since it's not the role of the FFT Node to bother with interpolation and for DRY purposes. 

> Document relative parameter for Bands 

OK ! 👍

jonathanjfshaw commented:
My understanding of fft dsp is very weak, so some of these may be things to document not change. Anything I learn from discussing, I will add to the docs.

> > Allow FFT to interpolate NaNs, so that a whole 20 second window is not discarded because of a single NaN

> NOK ! I prefer the user to use an interpolation node before the FFT if needed, since it's not the role of the FFT Node to bother with interpolation and for DRY purposes.

Agreed, and I will document. It's also important the the sample spacing is constant, so probably the  dejitter node should be used. I'll look into that.

> >Allow FFT nfft to vary dynamically dependent on input length

> Why would you want such a thing? To me, this node should be used as part of a Branch, with a Window node before to ensure that the chunks are of same length. Actually, we could even warn the user if this is not the case. The number of samples decides the frequency axis, and you don't want it to be changed.

If I put a Bands node after an FFT node, and my bands are wide, it's not obvious that it matters if the frequency axis varies in length.  

BUT if a window is in practice necessary anyway to provide a reliable amount of data, and a dejitter is necessary to provide constantly spaced samples, there's no possibility of the input length varying so what I proposed is pointless. I'll remove it.

>> Fix bug in FFT where one-sided (the default) was not working because pd.Series has no attribute 'real' in pd.apply

>>Don't bother converting dataframe to real numbers because https://docs.scipy.org/doc/numpy/reference/generated/numpy.fft.rfft.html#numpy.fft.rfft) says: "If the input a contains an imaginary part, it is silently discarded."

> I don't understand your point. Is there a related issue ?

I have not created an issue. When I tried a one-sided FFT I found it errored on spectral.py, line 112, FFT: `self.o.data = self.o.data.apply(lambda x: x.real)`
The error is because pd.Series has no attribute real, but also I think it might unnecessary anyway because the numpy rfft docs suggest to me that it will handle this for us.

jonathanjfshaw committed:
Remove interpolation and varying nfft

bertrandlalo commented:
Thanks @jonathanjfshaw for your valuable review and for updating the documentation based on our discussion. 
Agreed for your last point (ie. no conversion needed in case 'one sided'). 
Can you make the changes we just discussed?

jonathanjfshaw committed:
Remove amplitude, refactor sides

jonathanjfshaw committed:
Remove erroneous note about return_onesided

jonathanjfshaw commented:
Ok, changes made.

I wasn't fully certain after our discussions how you wanted to handle the relationship between nfft and input chunk length. I've suggested this:
```
        elif self._nfft != self.i.data.shape[0]:
            # np & scipy will pad/truncate the input as necessary, but this may not be desired.
            self.logger.debug("FFT input chunk has length {} but nfft is {}".format(self.i.data.shape[0], self._nfft))
```

Additionally, I refactored the handling of one/two-sided, as it was unnecessarily complex. Sorry if I'm suggesting too many changes in one PR.

The docs for `return_onesided` said:
```
                return_onesided (bool): If `True`, return a one-sided spectrum for real data.
                                              If `False` return a two-sided spectrum.
                                              (Note that for complex data, a two-sided spectrum is always returned.)
                                              Default: `True`.
```
This is copy-pasted from the scipy fft docs. I've removed the note as it did not seem true for us. If return_onesided was true we sent the input to nump.rfft which I expect will never return 2 sided.

Question: would it be appropriate to include a Branch node in this module that demonstrates the full combination of window, fft, and abs expression?

jonathanjfshaw commented:
I think this is ready now

bertrandlalo commented:
@jonathanjfshaw , thank you for your improvement. I think it would be great to add an example of a branch (eg. a welch, or else?) with the all combination. Do you want to do it? I'll be glad to review and think of how we could test it.

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -35,64 +36,67 @@ class FFT(Node):
-            self._nfft = int(self._nfft)
+        elif self._nfft != self.i.data.shape[0]:
+            # np & scipy will pad/truncate the input as necessary, but this may not be desired.
+            self.logger.debug("FFT input chunk has length {} but nfft is {}".format(self.i.data.shape[0], self._nfft))
```
I'd even log this as a warning, what do you think?

jonathanjfshaw replied:
I agree, because you've said that the chunk length not being constant is likely bad.

bertrandlalo commented on timeflux_dsp/nodes/spectral.py:
```diff
@@ -104,17 +108,21 @@ def update(self):
+            if 'rate' in self.i.meta:
+                self._rate = self.i.meta['rate']
+            else:
+                self.logger.debug("FFT using default sampling rate of 1; you should specify the rate explicitly or by meta.")
```
Same, let's make this log a warning!

jonathanjfshaw commented:
I agree, but I suggest it would keep the discussion better organised if we addressed that in a seperate issue, so I created #10

jonathanjfshaw committed:
Use warnings
